Angus/1272 save state

veggiesaurus opened the pull request:
**Description**
Closes #1272. Initial support for workspaces. The UI for saving/loading can certainly be improved (happy to delegate this to someone else as a separate issue), but this PR gets the core functionality ready for beta1. Once this is merged in, I will start adapting the controller to support this functionality as well

Companion PR of https://github.com/CARTAvis/carta-backend/pull/1240

Summary of functionality contained in workspace files:
- list of open files
- spatial/spectral/raster references
- regions
- zoom/pan
- channel settings (stokes/channel)
- render configs
- contour configs
- vector overlay configs

Workspaces can be loaded using the added menu items, or by using the `workspace` URL query parameter.

There are a number of features still to do, but they can be deferred until later:
- layout
- widgets
- analysis tools
- generated images (PV, moment)

Screencast of functionality :

https://user-images.githubusercontent.com/592504/220141292-22de0358-edc1-4ff1-8d2b-e7599c7d51b6.mp4




**Checklist**

For linked issues (if there are):
- [x] assignee and label added
- [x] ZenHub issue connection, board status, and estimate updated

For the pull request:
- [x] reviewers and assignee added
- [x] ZenHub estimate, milestone, and release (if needed) added
- [x] ~e2e test passing~ / corresponding fix added
- [x] changelog updated / ~~no changelog update needed~~
- [x] ~~protobuf updated to the latest dev commit~~ / no protobuf update needed
- [x] `BackendService` unchanged / ~~`BackendService` changed and corresponding ICD test fix added~~

veggiesaurus committed:
WIP loading example workspace

veggiesaurus committed:
Merge branch 'dev' into angus/1272_save_state

veggiesaurus committed:
WIP saving/loading workspace

veggiesaurus committed:
Merge branch 'dev' into angus/1272_save_state

veggiesaurus committed:
WIP saving/loading workspace

veggiesaurus committed:
Merge branch 'dev' into angus/1272_save_state

veggiesaurus committed:
WIP saving/loading workspace

veggiesaurus committed:
format fix

veggiesaurus committed:
import ordering

veggiesaurus committed:
change date heading

veggiesaurus committed:
Merge branch 'dev' into angus/1272_save_state

veggiesaurus committed:
only add useCallback when sensible

veggiesaurus committed:
load/save added to UI

veggiesaurus committed:
include a thumbnail when saving workspaces

veggiesaurus committed:
Merge branch 'dev' into angus/1272_save_state

veggiesaurus committed:
rollback config change

veggiesaurus commented on package.json:
```diff
@@ -113,6 +114,7 @@
                     "args": "none"
                 }
             ],
+            "object-curly-spacing": "off",
```
@YuHsuan-Hwang this solves the problem of eslint reformatting things when it rearranges imports

veggiesaurus committed:
rollback config change

veggiesaurus committed:
rollback config change

veggiesaurus commented on src/test/required_tiles.test.ts:
```diff
@@ -8,7 +8,6 @@ const Tile1024: Point2D = {x: 1024, y: 1024};
 const Tile2048: Point2D = {x: 2048, y: 2048};
 const Tile4096: Point2D = {x: 4096, y: 4096};
 const WideTile: Point2D = {x: 1024, y: 512};
-const TallTile: Point2D = {x: 512, y: 1024};
```
lint fix

veggiesaurus commented on src/registerServiceWorker.ts:
```diff
@@ -38,7 +38,7 @@ export default function register() {
                 // service worker/PWA documentation.
                 navigator.serviceWorker.ready.then(() => {
-                    console.log("This web app is being served cache-first by a service " + "worker. To learn more, visit https://goo.gl/SC7cgQ");
+                    console.log("This web app is being served cache-first by a service worker. To learn more, visit https://goo.gl/SC7cgQ");
```
lint fix

veggiesaurus committed:
package update

veggiesaurus committed:
thumbnail and workspace info added to workspace dialog

veggiesaurus committed:
regions and render configs added (basic functionality)

veggiesaurus committed:
skip cursor regions

veggiesaurus commented:
Regions and render configs now working, albeit limited capability:

https://user-images.githubusercontent.com/592504/214605619-a77aee07-93da-4e7f-a19c-4946522a0d40.mp4

veggiesaurus committed:
Merge branch 'dev' into angus/1272_save_state

veggiesaurus committed:
dev merge fixes

veggiesaurus committed:
adds support for contour, vector overlay, channel/stokes and pan/zoom

veggiesaurus committed:
initial support for fetching file info to check if workspace files are accessible

veggiesaurus committed:
remove spurious preferenceStore entry in contour and vector overlay configs

veggiesaurus committed:
write frontend version correctly in workspace output

veggiesaurus committed:
updated description

YuHsuan-Hwang commented on src/components/App/UIControllerComponent.tsx:
```diff
@@ -20,6 +20,8 @@ import {
     VectorOverlayDialogComponent
 } from "components/Dialogs";
 
+import {WorkspaceDialogComponent} from "../Dialogs/WorkspaceDialog/WorkspaceDialogComponent";
```
can be combined with "components/Dialogs"

YuHsuan-Hwang commented on src/components/Dialogs/WorkspaceDialog/WorkspaceInfoComponent.tsx:
```diff
@@ -0,0 +1,137 @@
+                const res = await AppStore.Instance.apiService.getWorkspace(name);
+                if (res) {
+                    setWorkspace(res);
+                    fetchFileInfo(res);
```
Do we need `await` here?

YuHsuan-Hwang commented on src/stores/HelpStore/HelpStore.ts:
```diff
@@ -9,6 +9,7 @@ export enum HelpType {
     FILE_BROWSER = "File Browser",
     FILE_INFO = "File Header",
     SAVE_LAYOUT = "Save Layout",
+    SAVE_WORKSPACE = "Save Layout",
```
typo: "Save Workspace"

YuHsuan-Hwang reviewed (changes requested):
Some initial comments:
1. The dialog is too large and can not be resized

2. When there are no workspace, two non ideal states are overlapped
<img width="400" alt="image" src="https://user-images.githubusercontent.com/43841102/222655946-c917d852-8016-47cc-a05c-81bcaa232837.png">

3. I think it would be convenient to be able to delete workspace. Currently we have `clearWorkspace` but not used by components.

4. I'm not sure if "close workspace" is useful if it's the same as "close all images".

5. The screenshot is small with devicePixelRatio = 2
<img width="400" alt="image" src="https://user-images.githubusercontent.com/43841102/222663789-de9a602f-dd14-4b3f-a20e-05776af76ebf.png">

Jordatious commented:
This is cool! I've also found the same issues mentioned above. 

When I try to save a workspace under the same name, it apparently saves, but duplicates the same name on a separate line in the workspace list, but is never accessible over the 1st occurrence. Obviously some features are missing, like mentioned above. Furthermore, locked regions become unlocked, and the frozen cursor position is lost.

veggiesaurus committed:
Merge branch 'dev' into angus/1272_save_state

veggiesaurus committed:
PR feedback fixes

veggiesaurus committed:
PR feedback fixes, dialog size and screenshot

veggiesaurus committed:
css adjustments to fix nonideal states and image size

veggiesaurus committed:
added workspace deletion and region locking to workspaces

veggiesaurus commented:
@YuHsuan-Hwang 

> Some initial comments:
> 
> 1. The dialog is too large and can not be resized
> 2. When there are no workspace, two non ideal states are overlapped

fixed

 
> 3. I think it would be convenient to be able to delete workspace. Currently we have `clearWorkspace` but not used by components.

I agree. I've added it as an action button next to open/save. Does this look ok? 
![image](https://user-images.githubusercontent.com/592504/230918537-dfe78ff7-92df-44d8-989c-dd1fa55e2107.png)
![image](https://user-images.githubusercontent.com/592504/230918598-46adb8ec-b53b-40a7-b6cc-42ec53755229.png)


> 4. I'm not sure if "close workspace" is useful if it's the same as "close all images".

We can remove that. @kswang1029 @Jordatious what do you think?

> 5. The screenshot is small with devicePixelRatio = 2

fixed

> 
> <img alt="image" width="400" src="https://user-images.githubusercontent.com/43841102/222663789-de9a602f-dd14-4b3f-a20e-05776af76ebf.png">


@Jordatious 

> When I try to save a workspace under the same name, it apparently saves, but duplicates the same name on a separate line in the workspace list, but is never accessible over the 1st occurrence.

I have a fairly good idea why this happens on the controller side. Shouldn't happen in the normal build though :innocent: 

> Furthermore, locked regions become unlocked, and the frozen cursor position is lost.

Locked regions added. I'm not entirely sure how to solve the frozen cursor position :thinking:  this might have to be a known issue for beta1

veggiesaurus commented:
@kswang1029 please re-test, I have addressed the comments above

kswang1029 commented:
@veggiesaurus overall this looks good!

- if we delete a workspace, the screenshot and its description at the right panel of the workspace dialog remain. This should become the state when no workspace file is selected.

- in the workspace dialog, we do not need the "close" button. We can just use the X button at the top-right corner of the dialog. This would also prevent confusion about what "close" means.

- File menu issues:
We better group the context as

```
-------------------
Open Image
Append Image
Save Image
Close Image
-------------------
Import Regions
Export Regions
Import Catalog
Export Image      >
-------------------
Open Workspace
Save Workspace
-------------------
Preferences
Server            >
-------------------
```

The "Close Workspace" option is not needed I think.

- Widgets menu issues
We are adopting title-cased style in the menu. Eg "Statistics Widget"

- If we load a workspace with images matched spatially and spectrally then if we load a workspace with images matched only spatially (single channel images), there will be a set of warning toasts (orange color) showing up saying for example

"Could not enable spectral matching (velocity system) of cosmos_spitzer24micron.fits to reference image member.uid___A001_X1284_X6c1._G31.41p0.31__sci.spw25.cube.I.pbcor_subimage.fits. No valid transform was found"

- do you think we can use the blueprintjs icon "control" or "applications" as the icon at the top-left corner of the workspace dialog? 

- it may be good to add some text info in the dialog if the selected workspace file has invalid images.

- it appears that the image viewer grid layout is not part of the workspace file. If we have a 3x3 grid with 9 images loaded as a workspace, when it is loaded back with a grid config as 2x2, the 9 images are loaded in 2x2 layout.

YuHsuan-Hwang commented:
@veggiesaurus The new changes are nice. Some other comments:

* Looks like the "Last modified" time changes randomly when I reopen the dialog

* I suggest we set the order of action buttons to be `Delete` `Open/Save`

* I would prefer the input column without the left and right margin
  <img src="https://user-images.githubusercontent.com/43841102/231089026-b9e69bcf-2692-4e2a-b953-487220eecfbe.png" width="400" />

* (good to have) It would be convenient if we can scroll the workspace info panel in x direction

Jordatious commented:
I'm happy to test this once those changes are made. Is this / will this be deployed to the IDIA carta-testing server?

veggiesaurus committed:
Merge branch 'dev' into angus/1272_save_state

veggiesaurus commented:
> 

I'll deploy it once things are merged into dev :+1:

veggiesaurus committed:
Merge branch 'dev' into angus/1272_save_state

veggiesaurus committed:
warns if a workspace contains gneerated files

veggiesaurus committed:
Merge branch 'dev' into angus/1272_save_state

veggiesaurus committed:
more PR feedback addressed

veggiesaurus committed:
fixes spatial matching bug

veggiesaurus committed:
reformatting

veggiesaurus commented:
@YuHsuan-Hwang @kswang1029 can you re-reivew please?

I've fixed the above issues, apart from:
- input group CSS (I can fix in a separate branch if needed, or do a quick update to this branch on Monday)
- Workspace info panel scroll and invalid information text (I think we need to think about how to communicate this to the user and how to still allow them to load files that _are_ valid).
- Image view layout (e.g. grid) needs more thought and discussion

I think the above can easily be fixed for 4.0 stable, as well as expanded support for workspaces.

kswang1029 commented:
@veggiesaurus I spotted a regression (not in dev) with e2e tests. 
To reproduce:
1. load an image
2. enable region list widget in the view
3. load a region file with multiple regions (eg 4 regions), we will see in the region list widget there are four regions with default names as Region 1, Region 2, etc.
4. click the image view and we will see those region names are updated to Region 5, Region 6, etc. In the console, we also see logs about updating region id. 


https://user-images.githubusercontent.com/20819712/235415884-98ce4220-10f4-4806-af20-63830e2b6113.mov

kswang1029 commented:
FYI. This branch introduced an xpath index shift. I have created a fix branch for e2e tests. Once the region id regression is fixed, I will try to ensure e2e tests work as usual. This is required for v4-beta package e2e testing.

veggiesaurus committed:
fixed regression with temporary region flag

veggiesaurus commented:
> @veggiesaurus I spotted a regression (not in dev) with e2e tests. To reproduce:
> 
> 1. load an image
> 2. enable region list widget in the view
> 3. load a region file with multiple regions (eg 4 regions), we will see in the region list widget there are four regions with default names as Region 1, Region 2, etc.
> 4. click the image view and we will see those region names are updated to Region 5, Region 6, etc. In the console, we also see logs about updating region id.

@kswang1029 fixed :innocent:

kswang1029 commented:
@veggiesaurus some issues:
1. if we want to save a workspace with a name colliding with existing workspaces, the existing one will be overwritten without any warning or confirmation. 
2. if the filename is short, the screenshot is displayed nicely. But if the filename is long, the screenshot seems to align with the width of the long filename, leading to a partial display of the screenshot. We need to manually enlarge the window for a better view (and UX). Wonder if it is possible to decouple the alignment (screenshot and text) here?
![Screen Shot 2023-05-02 at 09 09 43](https://user-images.githubusercontent.com/20819712/235558558-b4e655aa-46e8-4be0-87c2-6b54ae58bca9.png)
![Screen Shot 2023-05-02 at 09 10 05](https://user-images.githubusercontent.com/20819712/235558574-d38b20cc-dbdb-455e-9111-bd9b65810ef2.png)
3. In the above screenshot, we see the filename input field is not aligned with the upper two panels (it is narrower)
4. to be consistent, the delete button should be in red with a trash can icon (cf the region dialog).
![Screen Shot 2023-05-02 at 09 16 04](https://user-images.githubusercontent.com/20819712/235558960-cf6d98e0-ec00-4a8d-b17e-b70290e2fcfd.png)
5. It is possible to save a workspace with "whitespace" as a name. However, if we try to open it, when we select it in the workspace file browser, the frontend will crash.
![Screen Shot 2023-05-02 at 09 19 09](https://user-images.githubusercontent.com/20819712/235559282-682d3481-dbdf-4655-88a6-cb15eb586ced.png)
6. please disable spelling check in the filename input field.

YuHsuan-Hwang commented on src/stores/AppStore/AppStore.ts:
```diff
@@ -1999,6 +2047,269 @@ export class AppStore {
+        let hasTemporaryFiles = false;
+
+        for (const frame of this.frames) {
+            if (frame.frameInfo.fileId < 0) {
```
This needs to be consistent with `inMemoryImages` in `resumeSession`.

YuHsuan-Hwang reviewed (changes requested):
Just one change request. I think most of the remaining issues are related to CSS or design.

veggiesaurus committed:
PR feedback fixes

veggiesaurus committed:
disables spellcheck on input

veggiesaurus commented:
@YuHsuan-Hwang fixed

@kswang1029 fixed the following:
- whitespace characters in name (requires latest commit of workspaces backend)
- delete button intent and icon
- spellcheck disabled

The rest are CSS and minor, and can be done for stable

YuHsuan-Hwang commented:
This PR is ready for merge once the changelog is updated.

veggiesaurus committed:
handle annotation styles in workspace regions

veggiesaurus committed:
updates changelog

veggiesaurus committed:
Merge branch 'dev' into angus/1272_save_state

veggiesaurus committed:
fixed accidental commit of config

veggiesaurus committed:
updated protobuf

YuHsuan-Hwang committed:
Merge branch 'dev' into angus/1272_save_state
